packages/opentelemetry-instrumentation-anthropic/opentelemetry/instrumentation/anthropic/__init__.py

```diff
@@ -308,27 +345,23 @@ def _set_token_usage(
     if usage := response.get("usage"):
         prompt_tokens = usage.input_tokens
     else:
-        prompt_tokens = 0
-        if hasattr(anthropic, "count_tokens"):
-            if request.get("prompt"):
-                prompt_tokens = anthropic.count_tokens(request.get("prompt"))
-            elif request.get("messages"):
-                prompt_tokens = 0
-                for m in request.get("messages"):
-                    content = m.get("content")
-                    if isinstance(content, str):
-                        prompt_tokens += anthropic.count_tokens(content)
-                    elif isinstance(content, list):
-                        for item in content:
-                            # TODO: handle image tokens
-                            if isinstance(item, dict) and item.get("type") == "text":
-                                prompt_tokens += anthropic.count_tokens(
-                                    item.get("text", "")
-                                )
-
-    if token_histogram and type(prompt_tokens) is int and prompt_tokens >= 0:
+        prompt_tokens = count_prompt_tokens_from_request(anthropic, request)
+
+    if usage := response.get("usage"):
+        cache_read_tokens = dict(usage).get("cache_read_input_tokens", 0)
+    else:
+        cache_read_tokens = 0
+
+    if usage := response.get("usage"):
+        cache_creation_tokens = dict(usage).get("cache_creation_input_tokens", 0)
+    else:
+        cache_creation_tokens = 0
+
+    input_tokens = prompt_tokens + cache_read_tokens + cache_creation_tokens
```

Comment:
Aren't you double-calculating the number of input tokens like this? From my understanding `cache_read_tokens+cache_creation_tokens` should be exactly the number of tokens in the input. Or is it the case that either `prompt_tokens` is set (for non-cached requests) OR `cache_read_tokens+cache_creation_tokens`?

Reply:
Nope, Anthropic fills all three in. If it is a cache write, `cache_read_tokens == 0`. If it as a cache read, `cache_creation_tokens == 0`. `prompt_tokens` are tokens from uncached parts of the messages. For cache writes and reads is always about 3-4. I am assuming this is some control tokens or stop sequences or something. You can see some numbers I hard-coded in the tests [here](https://github.com/traceloop/openllmetry/blob/0cabaae2c86ff4d1a5957b04adbcb9108c03ddb4/packages/opentelemetry-instrumentation-anthropic/tests/test_completion.py#L624-L644).

For example, if I send two text blocks of sizes 1200 and 100 tokens in one message and only direct Anthropic to cache the first one, the usage will be:
- `{"cache_read_input_tokens": 0, "cache_creation_input_tokens": 1200, "input_tokens": 104, "output_tokens": ...}` for the first call, and
- `{"cache_read_input_tokens": 1200, "cache_creation_input_tokens": 0, "input_tokens": 104, "output_tokens": ...}` for the second

Reply:
My intuition is that we should keep the `input_tokens` constant across providers - so it should always be the number of tokens in the input - regardless if some of them were cached and some weren't

Reply:
Yes, except for one place where I accidentally added just uncached tokens. Fixed in the last commit now.